apm-agent-core/src/main/java/co/elastic/apm/agent/metrics/builtin/AgentOverheadMetrics.java

```diff
@@ -0,0 +1,442 @@
+        setThreadAllocatedMemoryEnabled = JmxUtils.getThreadMBeanMethod(threadBean, "setThreadAllocatedMemoryEnabled", boolean.class);
+        getThreadAllocatedBytes = JmxUtils.getThreadMBeanMethod(threadBean, "getThreadAllocatedBytes", long.class);
+
+        // See the documentation of J9 getProcessCpuTime for this special case:
+        // https://eclipse-openj9.github.io/openj9-docs/api/jdk8/jre/management/extension/com/ibm/lang/management/OperatingSystemMXBean.html#getProcessCpuTime--
+        if (JmxUtils.isIbmOperatingSystemMBean() && "true".equals(System.getProperty("com.ibm.lang.management.OperatingSystemMXBean.isCpuTime100ns"))) {
+            processCpuTimeScalingFactor = 100;
+        } else {
+            processCpuTimeScalingFactor = 1; //all other implementations return the time in nanoseconds
+        }
+    }
+
+    @Override
+    public void start(ElasticApmTracer tracer) throws Exception {
+        MetricRegistry metricRegistry = tracer.getMetricRegistry();
+        MetricsConfiguration config = tracer.getConfig(MetricsConfiguration.class);
+        bindTo(metricRegistry, config);
+    }
+
+    void bindTo(MetricRegistry metricRegistry, MetricsConfiguration config) {
+        boolean overheadMetricsEnabled = config.isOverheadMetricsEnabled();
+
+        cpuOverheadMetricEnabled = !metricRegistry.isDisabled(CPU_OVERHEAD_METRIC) && overheadMetricsEnabled;
+        cpuUsageMetricEnabled = !metricRegistry.isDisabled(CPU_USAGE_METRIC) && overheadMetricsEnabled;
+        allocationMetricEnabled = !metricRegistry.isDisabled(ALLOCATION_METRIC) && overheadMetricsEnabled;
+        threadCountMetricEnabled = !metricRegistry.isDisabled(THREAD_COUNT_METRIC) && overheadMetricsEnabled;
+
+        if (allocationMetricEnabled) {
+            boolean allocationMeasurementEnabled = enableThreadAllocationMeasurement();
+            if (!allocationMeasurementEnabled) {
+                allocationMetricEnabled = false;
+            }
+        }
+
+        if (cpuOverheadMetricEnabled || cpuUsageMetricEnabled) {
+            //ensure that process-cpu-time is supported, as it is required
+            boolean cpuTimeMeasurementEnabled = getProcessCpuTime() != NO_VALUE;
+            if (!cpuTimeMeasurementEnabled) {
+                logger.warn("Agent cpu overhead metrics can not be enabled: OperatingSystemMXBean.getProcessCpuTime() is not supported");
+            }
+            cpuTimeMeasurementEnabled = cpuTimeMeasurementEnabled && enableThreadCpuTimeMeasurement();
+            if (!cpuTimeMeasurementEnabled) {
+                cpuOverheadMetricEnabled = false;
+                cpuUsageMetricEnabled = false;
+            }
+        }
+
+        if (anyMetricEnabled()) {
+            //initialize values for first report
+            lastReportedProcessCpuTime = getProcessCpuTime();
+            ExecutorUtils.setThreadStartListener(this);
```

Comment:
How do you ensure that this is run before any agent threads have been created? Or do you just ignore threads created before this as not counted and they don't matter?

Reply:
> Or do you just ignore threads created before this as not counted and they don't matter

Partly this. To be more precise we ignore any cpu usage / allocations which have occured prior to this line.
However, we still count threads which were started before, we just start the counting at this point in time.

I thought that this was an okay limitation, because it is not really possible to capture the entire startup overhead:
E.g. for the overhead metrics, the configuration and logging needs to be initialized already. We could do the measurement right after that by relying on manual instantiation instead of `AbstractLifecycleListener`, but I thought that this is just not worth the introduced spaghetti by manual wiring.